File: movie_editor.py
import os
import subprocess
import argparse
from typing import Union

def time_to_seconds(time_str: str) -> int:
    """Convert time string in HH:MM:SS or MM:SS format to seconds."""
    parts = list(map(float, time_str.split(':')))
    if len(parts) == 3:
        h, m, s = parts
    elif len(parts) == 2:
        h, m, s = 0, parts[0], parts[1]
    else:
        raise ValueError("Invalid time format. Use HH:MM:SS or MM:SS.")
    return int(h * 3600 + m * 60 + s)

def check_file_exists(filename: str) -> bool:
    """Check if a file exists."""
    if not os.path.exists(filename):
        print(f"Error: The file {filename} does not exist.")
        return False
    return True

def handle_error(e: Exception) -> None:
    """Handle errors by printing the error message and listing files in the current directory."""
    print(f"Error: {e}")
    print("Files in the current directory:")
    for file in os.listdir('.'):
        print(file)

def run_ffmpeg_command(command: str) -> None:
    """Run an ffmpeg command."""
    try:
        subprocess.run(command, shell=True, check=True)
    except subprocess.CalledProcessError as e:
        handle_error(e)

def trim_video(filename: str, start_time: str, end_time: str, output_filename: str) -> None:
    """Trim a video from start time to end time."""
    if not check_file_exists(filename):
        return
    
    start_time_sec, end_time_sec = time_to_seconds(start_time), time_to_seconds(end_time)
    command = f"ffmpeg -i {filename} -ss {start_time_sec} -to {end_time_sec} -c copy {output_filename}"
    run_ffmpeg_command(command)
    print(f"Video trimmed and saved as {output_filename}")

def concatenate_videos(filename1: str, filename2: str, output_filename: str) -> None:
    """Concatenate two videos."""
    if not check_file_exists(filename1) or not check_file_exists(filename2):
        return
    
    with open("videos.txt", "w") as f:
        f.write(f"file '{filename1}'\n")
        f.write(f"file '{filename2}'\n")
    
    command = f"ffmpeg -f concat -safe 0 -i videos.txt -c copy {output_filename}"
    run_ffmpeg_command(command)
    os.remove("videos.txt")
    print(f"Videos concatenated and saved as {output_filename}")

def concatenate_videos_in_folder(folder: str, output_filename: str) -> None:
    """Concatenate all videos in a folder using ffmpeg."""
    if not os.path.isdir(folder):
        print(f"Error: The folder {folder} does not exist.")
        return
    
    video_files = [os.path.join(folder, f) for f in os.listdir(folder) if f.endswith(('.mp4', '.avi', '.mov'))]
    if not video_files:
        print(f"No video files found in the folder {folder}.")
        return

    with open("videos.txt", "w") as f:
        for video_file in video_files:
            f.write(f"file '{video_file}'\n")
    
    command = f"ffmpeg -f concat -safe 0 -i videos.txt -c copy {output_filename}"
    run_ffmpeg_command(command)
    os.remove("videos.txt")
    print(f"Videos concatenated and saved as {output_filename}")

def add_audio_to_video(video_filename: str, audio_filename: str, output_filename: str) -> None:
    """Add a WAV file to the sound of the video."""
    if not check_file_exists(video_filename) or not check_file_exists(audio_filename):
        return

    command = f"ffmpeg -i {video_filename} -i {audio_filename} -c:v copy -c:a aac {output_filename}"
    run_ffmpeg_command(command)
    print(f"Audio added and video saved as {output_filename}")

def fade_videos(filename1: str, filename2: str, output_filename: str, duration: int = 1) -> None:
    """Fade one video into another."""
    if not check_file_exists(filename1) or not check_file_exists(filename2):
        return
    
    temp_output1 = "temp_output1.mp4"
    temp_output2 = "temp_output2.mp4"
    
    command1 = f"ffmpeg -i {filename1} -vf 'fade=t=out:st=0:d={duration}' {temp_output1}"
    command2 = f"ffmpeg -i {filename2} -vf 'fade=t=in:st=0:d={duration}' {temp_output2}"
    
    run_ffmpeg_command(command1)
    run_ffmpeg_command(command2)
    
    concatenate_videos(temp_output1, temp_output2, output_filename)
    
    os.remove(temp_output1)
    os.remove(temp_output2)
    print(f"Videos faded and saved as {output_filename}")

def play_video(filename: str) -> None:
    """Play a video."""
    if not check_file_exists(filename):
        return
        
    command = f"ffplay {filename}"
    run_ffmpeg_command(command)

def play_all_videos_in_folder(folder: str) -> None:
    """Play all videos in a folder."""
    if not os.path.isdir(folder):
        print(f"Error: The folder {folder} does not exist.")
        return

    video_files = [os.path.join(folder, f) for f in os.listdir(folder) if f.endswith(('.mp4', '.avi', '.mov'))]
    if not video_files:
        print(f"No video files found in the folder {folder}.")
        return

    for video_file in video_files:
        print(f"Playing {video_file}...")
        play_video(video_file)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Trim, concatenate, fade, play videos, or add audio to a video.")
    subparsers = parser.add_subparsers(dest="command")

    trim_parser = subparsers.add_parser("trim", help="Trim a video from start time to end time.")
    trim_parser.add_argument("-f", "--filename", help="The name of the file to process.", required=True)
    trim_parser.add_argument("-s", "--start", help="The start time in HH:MM:SS or MM:SS format.", required=True)
    trim_parser.add_argument("-e", "--end", help="The end time in HH:MM:SS or MM:SS format.", required=True)
    trim_parser.add_argument("-o", "--output", help="The name of the output file.", required=True)

    concat_parser = subparsers.add_parser("concat", help="Concatenate two videos.")
    concat_parser.add_argument("-f1", "--filename1", help="The name of the first file to process.", required=True)
    concat_parser.add_argument("-f2", "--filename2", help="The name of the second file to process.", required=True)
    concat_parser.add_argument("-o", "--output", help="The name of the output file.", required=True)

    audio_parser = subparsers.add_parser("add_audio", help="Add a WAV file to the sound of the video.")
    audio_parser.add_argument("-v", "--video_filename", help="The name of the video file to process.", required=True)
    audio_parser.add_argument("-a", "--audio_filename", help="The name of the audio file to add.", required=True)
    audio_parser.add_argument("-o", "--output", help="The name of the output file.", required=True)

    fade_parser = subparsers.add_parser("fade", help="Fade one video into another.")
    fade_parser.add_argument("-f1", "--filename1", help="The name of the first file to process.", required=True)
    fade_parser.add_argument("-f2", "--filename2", help="The name of the second file to process.", required=True)
    fade_parser.add_argument("-o", "--output", help="The name of the output file.", required=True)
    fade_parser.add_argument("-d", "--duration", type=int, default=1, help="Duration of the fade effect in seconds.")

    play_parser = subparsers.add_parser("play", help="Play a video.")
    play_parser.add_argument("-f", "--filename", help="The name of the file to play.", required=True)

    play_all_parser = subparsers.add_parser("play_all_videos_in_folder", help="Play all videos in a folder.")
    play_all_parser.add_argument("-d", "--folder", help="The name of the folder to process.", required=True)

    concat_folder_parser = subparsers.add_parser("concatenate_videos_in_folder", help="Concatenate all videos in a folder.")
    concat_folder_parser.add_argument("-d", "--folder", help="The name of the folder to process.", required=True)
    concat_folder_parser.add_argument("-o", "--output", help="The name of the output file.", required=True)

    args = parser.parse_args()

    # Ensure ffmpeg is installed
    try:
        subprocess.run(["ffmpeg", "-version"], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    except subprocess.CalledProcessError:
        print("Error: ffmpeg is not installed. Please install ffmpeg to use this script.")
        exit(1)

    if args.command in ["trim", "t"]:
        trim_video(args.filename, args.start, args.end, args.output)
    elif args.command in ["concat", "c"]:
        concatenate_videos(args.filename1, args.filename2, args.output)
    elif args.command in ["add_audio", "aa"]:
        add_audio_to_video(args.video_filename, args.audio_filename, args.output)
    elif args.command in ["fade", "f"]:
        fade_videos(args.filename1, args.filename2, args.output, args.duration)
    elif args.command in ["play", "p"]:
        play_video(args.filename)
    elif args.command in ["play_all_videos_in_folder", "paf"]:
        play_all_videos_in_folder(args.folder)
    elif args.command in ["concatenate_videos_in_folder", "cvf"]:
        concatenate_videos_in_folder(args.folder, args.output)

File: movieEditor.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

func timeToSeconds(timeStr string) (int, error) {
	fmt.Printf("Converting time string %s to seconds\n", timeStr)
	parts := strings.Split(timeStr, ":")
	if len(parts) < 2 || len(parts) > 3 {
		return 0, fmt.Errorf("invalid time format. Use HH:MM:SS or MM:SS")
	}

	var h, m, s int
	var err error

	if len(parts) == 3 {
		h, err = strconv.Atoi(parts[0])
		if err != nil {
			return 0, err
		}
		m, err = strconv.Atoi(parts[1])
		if err != nil {
			return 0, err
		}
		s, err = strconv.Atoi(parts[2])
		if err != nil {
			return 0, err
		}
	} else {
		m, err = strconv.Atoi(parts[0])
		if err != nil {
			return 0, err
		}
		s, err = strconv.Atoi(parts[1])
		if err != nil {
			return 0, err
		}
	}

	seconds := h*3600 + m*60 + s
	fmt.Printf("Converted time string %s to %d seconds\n", timeStr, seconds)
	return seconds, nil
}

func checkFileExists(filename string) bool {
	fmt.Printf("Checking if file %s exists\n", filename)
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		fmt.Printf("Error: The file %s does not exist.\n", filename)
		return false
	}
	fmt.Printf("File %s exists\n", filename)
	return true
}

func handleError(err error) {
	fmt.Printf("Error: %v\n", err)
	fmt.Println("Files in the current directory:")
	files, _ := os.ReadDir(".")
	for _, file := range files {
		fmt.Println(file.Name())
	}
}

func runFFmpegCommand(command string) {
	fmt.Printf("Running command: %s\n", command)
	cmd := exec.Command("sh", "-c", command)
	err := cmd.Run()
	if err != nil {
		handleError(err)
	}
}

func createFileList(filenames []string) (string, error) {
	file, err := os.Create("videos.txt")
	if err != nil {
		return "", err
	}
	defer file.Close()

	for _, filename := range filenames {
		file.WriteString(fmt.Sprintf("file '%s'\n", filename))
	}

	return "videos.txt", nil
}

func trimVideo(filename, startTime, endTime, outputFilename string) {
	fmt.Printf("Trimming video %s from %s to %s, outputting to %s\n", filename, startTime, endTime, outputFilename)
	if !checkFileExists(filename) {
		return
	}

	startTimeSec, err := timeToSeconds(startTime)
	if err != nil {
		handleError(err)
		return
	}
	endTimeSec, err := timeToSeconds(endTime)
	if err != nil {
		handleError(err)
		return
	}

	command := fmt.Sprintf("ffmpeg -i %s -ss %d -to %d -c copy %s", filename, startTimeSec, endTimeSec, outputFilename)
	runFFmpegCommand(command)
	fmt.Printf("Video trimmed and saved as %s\n", outputFilename)
}

func concatenateVideos(filenames []string, outputFilename string) {
	fmt.Printf("Concatenating videos %v, outputting to %s\n", filenames, outputFilename)
	for _, filename := range filenames {
		if !checkFileExists(filename) {
			return
		}
	}

	fileList, err := createFileList(filenames)
	if err != nil {
		handleError(err)
		return
	}
	defer os.Remove(fileList)

	command := fmt.Sprintf("ffmpeg -f concat -safe 0 -i %s -c copy %s", fileList, outputFilename)
	runFFmpegCommand(command)
	fmt.Printf("Videos concatenated and saved as %s\n", outputFilename)
}

func concatenateVideosInFolder(folder, outputFilename string) {
	fmt.Printf("Concatenating all videos in folder %s, outputting to %s\n", folder, outputFilename)
	if _, err := os.Stat(folder); os.IsNotExist(err) {
		fmt.Printf("Error: The folder %s does not exist.\n", folder)
		return
	}

	files, err := os.ReadDir(folder)
	if err != nil {
		handleError(err)
		return
	}

	var videoFiles []string
	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".mp4") || strings.HasSuffix(file.Name(), ".avi") || strings.HasSuffix(file.Name(), ".mov") {
			videoFiles = append(videoFiles, filepath.Join(folder, file.Name()))
		}
	}

	if len(videoFiles) == 0 {
		fmt.Printf("No video files found in the folder %s.\n", folder)
		return
	}

	concatenateVideos(videoFiles, outputFilename)
}

func addAudioToVideo(videoFilename, audioFilename, outputFilename string) {
	fmt.Printf("Adding audio %s to video %s, outputting to %s\n", audioFilename, videoFilename, outputFilename)
	if !checkFileExists(videoFilename) || !checkFileExists(audioFilename) {
		return
	}

	command := fmt.Sprintf("ffmpeg -i %s -i %s -c:v copy -c:a aac %s", videoFilename, audioFilename, outputFilename)
	runFFmpegCommand(command)
	fmt.Printf("Audio added and video saved as %s\n", outputFilename)
}

func fadeVideos(filename1, filename2, outputFilename string, duration int) {
	fmt.Printf("Fading videos %s and %s with duration %d, outputting to %s\n", filename1, filename2, duration, outputFilename)
	if !checkFileExists(filename1) || !checkFileExists(filename2) {
		return
	}

	tempOutput1 := "temp_output1.mp4"
	tempOutput2 := "temp_output2.mp4"

	command1 := fmt.Sprintf("ffmpeg -i %s -vf 'fade=t=out:st=0:d=%d' %s", filename1, duration, tempOutput1)
	command2 := fmt.Sprintf("ffmpeg -i %s -vf 'fade=t=in:st=0:d=%d' %s", filename2, duration, tempOutput2)

	runFFmpegCommand(command1)
	runFFmpegCommand(command2)

	concatenateVideos([]string{tempOutput1, tempOutput2}, outputFilename)

	os.Remove(tempOutput1)
	os.Remove(tempOutput2)
	fmt.Printf("Videos faded and saved as %s\n", outputFilename)
}

func playVideo(filename string) {
	fmt.Printf("Playing video %s\n", filename)
	if !checkFileExists(filename) {
		return
	}

	command := fmt.Sprintf("ffplay %s", filename)
	runFFmpegCommand(command)
}

func playAllVideosInFolder(folder string) {
	fmt.Printf("Playing all videos in folder %s\n", folder)
	if _, err := os.Stat(folder); os.IsNotExist(err) {
		fmt.Printf("Error: The folder %s does not exist.\n", folder)
		return
	}

	files, err := os.ReadDir(folder)
	if err != nil {
		handleError(err)
		return
	}

	var videoFiles []string
	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".mp4") || strings.HasSuffix(file.Name(), ".avi") || strings.HasSuffix(file.Name(), ".mov") {
			videoFiles = append(videoFiles, filepath.Join(folder, file.Name()))
		}
	}

	if len(videoFiles) == 0 {
		fmt.Printf("No video files found in the folder %s.\n", folder)
		return
	}

	for _, videoFile := range videoFiles {
		fmt.Printf("Playing %s...\n", videoFile)
		playVideo(videoFile)
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run movieEditor.go <command> [<args>]")
		return
	}

	command := os.Args[1]
	fmt.Printf("Executing command: %s\n", command)

	switch command {
	case "trim", "-t", "trim_video", "--trim":
		// Example: go run movieEditor.go trim input.mp4 00:00:10 00:00:20 output.mp4
		if len(os.Args) != 6 {
			fmt.Println("Usage: go run movieEditor.go trim <filename> <start> <end> <output>")
			return
		}
		trimVideo(os.Args[2], os.Args[3], os.Args[4], os.Args[5])
	case "concat", "-c", "concatenate", "--concat", "c":
		// Example: go run movieEditor.go concat input1.mp4 input2.mp4 output.mp4
		if len(os.Args) != 5 {
			fmt.Println("Usage: go run movieEditor.go concat <filename1> <filename2> <output>")
			return
		}
		concatenateVideos([]string{os.Args[2], os.Args[3]}, os.Args[4])
	case "concat_folder", "-cf", "concat_folder_videos", "--concat_folder", "cf":
		// Example: go run movieEditor.go concat_folder videos_folder output.mp4
		if len(os.Args) != 4 {
			fmt.Println("Usage: go run movieEditor.go concat_folder <folder> <output>")
			return
		}
		concatenateVideosInFolder(os.Args[2], os.Args[3])
	case "add_audio", "-a", "add_audio_to_video", "--add_audio":
		// Example: go run movieEditor.go add_audio input.mp4 audio.mp3 output.mp4
		if len(os.Args) != 5 {
			fmt.Println("Usage: go run movieEditor.go add_audio <video_filename> <audio_filename> <output>")
			return
		}
		addAudioToVideo(os.Args[2], os.Args[3], os.Args[4])
	case "fade", "-f", "fade_videos", "--fade":
		// Example: go run movieEditor.go fade input1.mp4 input2.mp4 output.mp4 5
		if len(os.Args) != 6 {
			fmt.Println("Usage: go run movieEditor.go fade <filename1> <filename2> <output> <duration>")
			return
		}
		duration, err := strconv.Atoi(os.Args[5])
		if err != nil {
			handleError(err)
			return
		}
		fadeVideos(os.Args[2], os.Args[3], os.Args[4], duration)
	case "play", "-p", "play_video", "--play", "p":
		// Example: go run movieEditor.go play input.mp4
		if len(os.Args) != 3 {
			fmt.Println("Usage: go run movieEditor.go play <filename>")
			return
		}
		playVideo(os.Args[2])
	case "play_all", "-pa", "play_all_videos", "--play_all":
		// Example: go run movieEditor.go play_all videos_folder
		if len(os.Args) != 3 {
			fmt.Println("Usage: go run movieEditor.go play_all <folder>")
			return
		}
		playAllVideosInFolder(os.Args[2])
	default:
		fmt.Println("Usage: go run movieEditor.go <command> [<args>]")
		fmt.Println("Commands:")
		fmt.Println("  trim, -t, trim_video, --trim <filename> <start> <end> <output> - Trim a video")
		fmt.Println("  concat, -c, concatenate, --concat <filename1> <filename2> <output> - Concatenate two videos")
		fmt.Println("  concat_folder, -cf, concat_folder_videos, --concat_folder <folder> <output> - Concatenate all videos in a folder")
		fmt.Println("  add_audio, -a, add_audio_to_video, --add_audio <video_filename> <audio_filename> <output> - Add audio to a video")
		fmt.Println("  fade, -f, fade_videos, --fade <filename1> <filename2> <output> <duration> - Fade between two videos")
		fmt.Println("  play, -p, play_video, --play <filename> - Play a video")
		fmt.Println("  play_all, -pa, play_all_videos, --play_all <folder> - Play all videos in a folder")
	}
}
